hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java

```diff
@@ -648,67 +619,65 @@ public long countByInstalledDistributionSet(final long distId) {
     }
 
     @Override
-    public Page<Target> findByTargetFilterQueryAndNonDS(final Pageable pageRequest, final long distributionSetId,
-            final String targetFilterQuery) {
-        throwEntityNotFoundIfDsDoesNotExist(distributionSetId);
-
-        final Specification<JpaTarget> spec = RSQLUtility.parse(targetFilterQuery, TargetFields.class,
-                virtualPropertyReplacer, database);
-
-        return findTargetsBySpec(
-                (root, cq,
-                        cb) -> cb.and(spec.toPredicate(root, cq, cb), TargetSpecifications
-                                .hasNotDistributionSetInActions(distributionSetId).toPredicate(root, cq, cb)),
-                pageRequest);
-
+    public Page<Target> findByQueryAndNotInRolloutGroups(final Pageable pageRequest, final Collection<Long> groupIds,
```

Comment:
could be refactored:
```
    @Override
    public Page<Target> findByQueryAndNotInRolloutGroups(final Pageable pageRequest, final Collection<Long> groups,
            final String rsqlParam) {

        final List<Target> targets = new LinkedList<>();
        final int maxSizeInPage = 500;

        IntStream.iterate(0, i -> i++)
                .mapToObj(currentPageNumber -> PageRequest.of(currentPageNumber, maxSizeInPage, ORDERBY_CONTROLLERID))
                .map(request -> targetQueryExecutionManagement.findByQuery(request, rsqlParam))
                .peek(page -> targets.addAll(targetsNotInGroups(page.getContent(), pageRequest.getPageSize() - targets.size(), groups)))
                .allMatch(Page::hasContent);

        return new PageImpl<>(targets, pageRequest, countByQueryAndNotInRolloutGroups(groups, rsqlParam));
    }

    private Collection<Target> targetsNotInGroups(final List<? extends Target> targets, final int maxSize,
            final Collection<Long> groups) {

        if(targets.isEmpty()) {
            return Collections.emptyList();
        }
        
        final Set<String> controllerIds = targets.stream().map(Target::getControllerId).collect(Collectors.toSet());
        final Specification<JpaTarget> specification = hasControllerId(controllerIds).and(isNotInRolloutGroups(groups));
        return targetRepository.findAll(specification).stream().limit(maxSize).collect(Collectors.toList());
    }
```
the `allMatch` is a shortcircuiting method, meaning, the loop execution will exit if the page has no content

Reply:
You could also remove the call to `countByQueryAndNotInRolloutGroups()` because basically it's making the pageable calls twice here. What we could do is refactor the current method to fetch all elements, then returning only the desired limit:

```
    @Override
    public Page<Target> findByQueryAndNotInRolloutGroups(final Pageable pageRequest, final Collection<Long> groups,
            final String rsqlParam) {

        final List<Target> foundTargets = new LinkedList<>();
        final int maxSizeInPage = 500;

        IntStream.iterate(0, i -> i++)
                .mapToObj(currentPageNumber -> PageRequest.of(currentPageNumber, maxSizeInPage, ORDERBY_CONTROLLERID))
                .map(request -> targetQueryExecutionManagement.findByQuery(request, rsqlParam))
                .peek(page -> foundTargets.addAll(targetsNotInGroups(page.getContent(), groups)))
                .allMatch(Page::hasContent);

        final List<Target> targets = foundTargets.stream().limit(pageRequest.getPageSize()).collect(Collectors.toList());
        return new PageImpl<>(targets, pageRequest, foundTargets.size());
    }
```

Reply:
- I do not see an advantage for using streams in your first suggestion - also I am not sure if the suggested implementation picks the right requested page-number or just fills up to "page-size".
- I spend some time to test and debug the current implemetation so I prefere to leave it for the moment
- It is a valid point with the double page-fetch:
  - I thought about a "parallel fetch of all targets" and/or
  - a specilized method like "findControllerIdsByQuery" which would reduce messege size in the remote case.

Both options would produce some effort as well - so can we leave it for now as it is and I do an optimization of the rollout start in a seperate PR?

Reply:
About the `@Transactional`:
`JpaRolloutManagement` does some custom Transaction-management - so I prefere not to mix-in some new Spring Transactional aspect here but leave it as it was before.

Reply:
>     * I do not see an advantage for using streams in your first suggestion - also I am not sure if the suggested implementation picks the right requested page-number or just fills up to "page-size".
> 
>     * I spend some time to test and debug the current implemetation so I prefere to leave it for the moment
> 
>     * It is a valid point with the double page-fetch:
>       
>       * I thought about a "parallel fetch of all targets" and/or
>       * a specilized method like "findControllerIdsByQuery" which would reduce messege size in the remote case.
> 
> 
> Both options would produce some effort as well - so can we leave it for now as it is and I do an optimization of the rollout start in a seperate PR?



Firstly, let's agree that either way this piece of code is ugly and has to be changed, I understand that changing it inside of this PR would be a lot of effort, but sooner or later we have to do it.
* Personally speaking, I find the approach with Streams more concise and readable. 
* In the current implementation the requested page number is also not taken in consideration.
* You can easily fix that with the stream by providing a skip operation:
```
// assuming that paging starts with 0, you could also try pageRequest.offset, not sure if produces the same output
final int offset = pageRequest.getPageNumber() * pageRequest.getPageSize();
final List<Target> targets = foundTargets.stream().skip(offset).limit(pageRequest.getPageSize()).collect(Collectors.toList());
```
* To get rid of the double fetch operation it's quite easy to implement with the streams as shown in my examples.
* We have to test this anyway, so I guess testing efforts are not a role player here

Reply:
As you mentioned also the old code didn't implemented the "paging" properly - but it was working (in the start-rollout) without issues. So I decided to get rid of the page-request as it was never really used by the calling code and so we do not need to know the "total-amount" of all targets anymore. 

I still had issues with the stream aproach and I do not like the fact that `allMatch` and the like is called once per element in the stream (meaning 500 targets found allMatch is called 500 times). Also the use of `peek` the get the results is IMHO not way streams are meant.

I hope the new implementation findes some mercy in the review-community ;-)

Reply:
The `allMatch` method is called per each request, not per each target, it's just a call to check if the `Page::hasContent` boolean is true or false, which I think is pretty cheap.
I agree with you about the `peek` concern, it is really not the way to do streams, but since this stream is coming from an external provider (we query an external service to get the targets), it is not possible to determine when to stop the stream unless we make this check ourselves with the `Page::hasContent` and the `allMatch`.
What kind of issues do you get when you try this apporach?

Reply:
True with the `allMatch` -- I think I had sth. like an endless-loop and the debugger was standing on this line...
If you also do not feele too good about the `peek` here it is maybe an indicator about that streams are not always the optimal aproach.

The new code looks like this:
```java
    public List<Target> findByQueryAndNotInRolloutGroups(int limit, final Collection<Long> groupIds,
            final String query) {
        final List<Target> foundTargets = new LinkedList<>();
        Page<Target> queryResult = null;

        while ((queryResult == null || queryResult.hasNext()) && foundTargets.size() < limit) {
            queryResult = targetQueryExecutionManagement.findByQuery(nextPageable(queryResult), query);
            foundTargets.addAll(findTargetsNotInGroup(queryResult, groupIds));
        }
        if (foundTargets.size() > limit) {
            return foundTargets.subList(0, limit);
        }
        return foundTargets;
    }
```

... and by thinking, that this part of code is a candidate to be removed completely, I tend to say "it's good enough".